CMakeLists.txt

```diff
@@ -2,6 +2,9 @@ cmake_minimum_required (VERSION 2.6)
 
 project (OPENCL_ICD_LOADER)
 
+option (BUILD_SHARED_LIBS "Build shared libs" ON)
```

Comment:
I don't see where this option is used, so it looks like it's only possible to build a static library with these changes.

What is the usage model for the static library?  The nice thing about a dynamic / shared library is that it can be updated asynchronously from applications.  With a static library there's a very real risk that applications may stop working - see the recent HKR changes in this project, as an example.

Reply:
The `BUILD_SHARED_LIBS` is the CMake "magic" option to build libraries as shared, OFF by default. So this change make it ON by default to keep the same behavior as before. But it also allows to build this as a static library if someone wants to.

"updated asynchronously"? What do you mean? Windows Update?

Reply:
The ICD loader is primarily designed to be a shared library (OpenCl.dll or libOpenCL.so). It is meant to be shipped either as a part of a vendor driver installation (Windows, Linux) or as part of a platform-specific package that contains the loader (Linux distros). What @bashbaug means is that the shared library can be updated by updating such installations (the display driver on Windows, for example), independent of whether the application is updated or not. IOW the update of the shared library is "asynchronous" w.r.t. to the application update.

@chfast, can you explain your use case where using a statically linked ICD loader is more useful compared to using the shared library version? There are disadvantages to doing that as @bashbaug mentioned above.

Reply:
Yes, I'm using this build as a static library in [ethminer](https://github.com/ethereum-mining/ethminer). In the and I have a single binary for every OS. And it will not crash in case OpenCL is not installed in the system.

Reply:
I see that the problem you are trying to solve applies to shared libraries in general, not just the OpenCL ICD loader. Any application that links to a shared library needs the library to be installed on the system or it will fail to launch, which is normal user expectation.

Reply:
I wish to build ffmpeg as a single STATIC .exe file including the ICD loader.
I can do this with some juggling, .  When it runs it appears to look for the manufacturer installed  OpenCL.DLL and load that.
That is one reason why the ICD loader needs to be built static, baked into in the ffmpeg exe - to avoid dll hell.

Reply:
Can you explain me how do you see the OpenCL loader to be shipped to Windows as a shared library? Via Windows Update?

Reply:
@chfast I am assuming you are asking me and not @hydra3333.

The OpenCL loader library (OpenCL.dll) on Windows is installed by OpenCL driver installer from a vendor like Intel or NVIDIA. Windows Update may install or update OpenCL.dll as part of installing or updating the vendor's driver. Windows Update by itself is not the primary mechanism to ship OpenCL.dll.

Reply:
@hydra3333 Can you clarify why dynamically loading OpenCL.dll does not solve the problem for you?

If I got you correctly, with dynamic loading your executable will not crash even when OpenCL.dll is not present. The executable will also benefit from future updates to the ICD loader when the DLL is updated. Sounds like a win-win to me instead of linking statically.

Reply:
Those install vendor's implementations of OpenCL, not a generic loader, right?

Anyway, this discussion is quite a waste of time. Are these code changes an issue or not?

Reply:
Vendor installers install two things: The OpenCL ICD loader (OpenCL.dll) and the OpenCL vendor implementation (<vendor-specific-name>.dll). The former is expected to be built using ICD loader source from Khronos (which is this repository, where we are talking right now).

So the OpenCL.dll is generic (vendor-agnostic) even if it comes from a vendor installer. The vendor-specific implementation resides in the second DLL and it gets loaded dynamically by the ICD loader.

@chfast I understand your frustration with the delay in accepting this change but please bear with us for now. The OpenCL working group appreciates your patch and we are actively discussing it within the group.

In this discussion with you and others we are trying to make sure we understand the rationale and use cases for building and using the ICD loader as a static library in shipped applications, while we know there are drawbacks to that approach compared to using a shared library.